### PR TITLE
Middleware

### PR DESCRIPTION
**Pull request by benwilson512**

### CODE REVIEW

**Comment by benwilson512** [comment]
TODO:
- [ ] Move pre resolution and post resolution hooks to phases
- [ ] docs
 -[ ] optimizations

**Comment by bruce** [comment]
Remember to point this at `next` -- my logging work is in now, once this in in we should be good for beta (just need to release both packages)

File: lib/absinthe/middleware.ex [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  struct and will also need to return one such struct.
+
+  On that struct there is a `context` key which holds the absinthe context. This
+  is generally where things like the current user are placed. For more information
+  on how the current user ends up in the context please see our full authentication
+  guide on the website.
+
+  Our `def call` function simply checks the context to see if there is a current
+  user. If there is, we pass the resolution onward. If there is not, we update
+  the resolution state to `:halt` and place an error result.
```
Comment (benwilson512) [inline lib/absinthe/middleware.ex]: - [ ] FIXME: This should either not mention the state, or fix it to `:resolved`

Comment (dpehrson) [inline lib/absinthe/middleware.ex]: Here it says that the resolution state is updated to `:halt`. Line 29 says `:error`. Is the halting a byproduct state of `put_result` an `:error` or should one of these be updated?

File: lib/absinthe/middleware.ex [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    field :lookup_user, :user do
+      resolve fn _, _ ->
+        {:ok, %{name: "Bob"}}
+      end
+    end
+  end
+
+  def set_middleware(field, object) do
+    # what is field?
+    # what is object?
```
Comment (dpehrson) [inline lib/absinthe/middleware.ex]: Are these two "what is X?" comments meant for the end user? As a reader I'm not sure what these comments are supposed to teach me.

File: lib/absinthe/middleware.ex [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  pattern match on an empty list. Generating the camelized key is a simple matter
+  of camelizing the field identifier.
+
+  Next we need to build a list that defines what middleware we want to use. The
+  form we're using is `{{MODULE, :function_to_call}, options_of_middleware}`. For
+  our purposes we're simply going to use a function in the schema module itself
+  `get_camelized_key`.
+
+  Like all middleware functions, it takes a resolution struct, and options. The
+  options is the caemlized key we generated. We get the camelized string from
```
Comment (dpehrson) [inline lib/absinthe/middleware.ex]: Typo: `caemlized` => `camelized`

File: lib/absinthe/middleware.ex [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+      field
+      |> Map.update!(:middleware, [MyApp.Web.Authentication | &1])
+    end
+    def set_middleware(field, _object) do
+      field
+    end
+  end
+  ```
+
+  ## TL;DR:
```
Comment (dpehrson) [inline lib/absinthe/middleware.ex]: Perhaps a better header here? Depends on the tone you want to set in the documentation.

File: lib/absinthe/middleware.ex [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      field :private_field, :string do
+        resolve fn _, _ ->
+          {:ok, "this can only be viewed if authenticated"}
+        end
+      end
+    end
+
+    def set_middleware(field, %Absinthe.Type.Object{identifier: :query}) do
+      field
+      |> Map.update!(:middleware, [MyApp.Web.Authentication | &1])
```
Comment (dpehrson) [inline lib/absinthe/middleware.ex]: For documentation examples, is there too much going on here language-wise or should we expect all readers of absinthe documentation to have a minimum level of familiarity with some of the fun language features of elixir like `[Module | &1]`?

File: test/lib/absinthe/schema/notation_test.exs [hunk 6]
```diff
@@ -18,6 +18,24 @@ defmodule Absinthe.Schema.NotationTest do
 
       assert [:email, :name] = Foo.__absinthe_type__(:bar).fields |> Map.keys |> Enum.sort
     end
+    #
```
Comment (dpehrson) [inline test/lib/absinthe/schema/notation_test.exs]: Just tagging commented code. Finish or delete before merge?

File: lib/absinthe/middleware.ex [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+  By specifying `@behaviour Absinthe.Middleware` the compiler will ensure that
+  we provide a `def call` callback. This function takes an `%Absinthe.Resolution{}`
+  struct and will also need to return one such struct.
+
+  On that struct there is a `context` key which holds the absinthe context. This
+  is generally where things like the current user are placed. For more information
+  on how the current user ends up in the context please see our full authentication
+  guide on the website.
+
+  Our `def call` function simply checks the context to see if there is a current
```
Comment (scrogson) [inline lib/absinthe/middleware.ex]: Maybe "Our `call/2` function..."

File: lib/absinthe/middleware.ex (hunk 5 above)
Comment (benwilson512) [inline lib/absinthe/middleware.ex]: This is a good question. At the moment I'm largely just pushing any and all middleware related content into this moduledoc. The plan prior to the final 1.3 release is to break out the simpler, elixir newbie friendly stuff into a Middleware Guide, break out some of the specific use cases (authorization, ecto changeset error handling) into specific guides, and then leave this module doc to be a more advanced look at the mechanics of middleware.

**Pull request merged**